lib/active_merchant/billing/gateways/paypal_checkout.rb

```diff
@@ -0,0 +1,298 @@
+        post[key] = {}
+        post[key][:full_name] = options[:full_name]
+        skip_empty(post[key], :full_name)
+      end
+
+      def add_shipping_address(address, obj_hsh, key = :address)
+        requires!(address, :admin_area_2, :postal_code, :country_code )
+
+        obj_hsh[key] = {}
+        obj_hsh[key][:address_line_1] = address[:address_line_1] unless address[:address_line_1].nil?
+        obj_hsh[key][:address_line_2] = address[:address_line_2] unless address[:address_line_2].nil?
+        obj_hsh[key][:admin_area_1]   = address[:admin_area_1] unless address[:admin_area_1].nil?
+        obj_hsh[key][:admin_area_2]   = address[:admin_area_2]
+        obj_hsh[key][:postal_code]    = address[:postal_code]
+        obj_hsh[key][:country_code]   = address[:country_code]
+        obj_hsh
+      end
+
+      def add_billing_address(address, obj_hsh)
+        requires!(address, :country_code )
+
+        obj_hsh[:billing_address] = {}
+        obj_hsh[:billing_address][:address_line_1] = address[:address_line_1] unless address[:address_line_1].nil?
+        obj_hsh[:billing_address][:admin_area_1]   = address[:admin_area_1] unless address[:admin_area_1].nil?
+        obj_hsh[:billing_address][:admin_area_2]   = address[:admin_area_2] unless address[:admin_area_2].nil?
+        obj_hsh[:billing_address][:postal_code]    = address[:postal_code] unless address[:postal_code].nil?
+        obj_hsh[:billing_address][:country_code]   = address[:country_code] unless address[:country_code].nil?
+        obj_hsh
+      end
+
+      def add_payment_source(source, post)
+        post[:payment_source] = {}
+        add_customer_card(source[:card], post[:payment_source]) unless source[:card].nil?
+        add_token(source[:token], post[:payment_source]) unless source[:token].nil?
+        skip_empty(post, :payment_source)
+      end
+
+      def add_customer_card(card_details, post)
+        requires!(card_details, :number, :expiry, :name, :security_code)
+
+        post[:card] = {}
+        post[:card][:name]          = card_details[:name]
+        post[:card][:number]        = card_details[:number]
+        post[:card][:expiry]        = card_details[:expiry]
+        post[:card][:security_code] = card_details[:security_code]
+        add_billing_address(card_details[:billing_address], post) unless card_details[:billing_address].nil?
+        verify_card(post[:card])
+        post
+      end
+
+      def verify_card(card)
+        defaults = {
+            number: card[:number],
+            first_name: card[:name],
+            last_name: card[:name],
+            verification_value: card[:security_code],
+            month: card[:expiry].split("-")[1].to_i,
+            year: card[:expiry].split("-")[0].to_i
+        }
+        @visa_card = ActiveMerchant::Billing::CreditCard.new(defaults)
+        raise "Invalid Credit Card Format. Message: Missing #{@visa_card.validate}" unless @visa_card.validate.empty?

```

Comment:
Why is this called visa card?

Reply:
since we never set the vallidate field, will it even validate? what does this provide us?

Reply:
The validate function is available in the CreditCard class here - https://github.com/inaitech/active_merchant/blob/master/lib/active_merchant/billing/credit_card.rb#L306 which validates the card number. Added this validation here to check the given card details.
I will rename the reference though.